Design document for adding roles to manual judgment stage.

sanopsmx opened the pull request:
# Select roles that can execute manual judgement stage

| | |
|-|-|
| **Status**     | _**Proposed**, Accepted, Implemented, Obsolete_ |
| **RFC #**      | https://github.com/spinnaker/deck/pull/8565 , https://github.com/spinnaker/orca/pull/3947
| **Author(s)**  | Sanjeev Thatiparthi (@sanopsmx) (https://github.com/sanopsmx)
| **SIG / WG**   | sig-ux

## Overview

 Display all the roles of the user added in the application in the manual judgment stage.If there are 2 groups(dev,qa)
 and qa group has the permission to deploy.If there is a manual judgment stage before the deploy stage.If there are
 no groups attached to the manual judgment stage, then dev group can continue with the manual judgment stage and get
 failed at the deploy stage as only qa group has the permission to deploy. This is the reason we are displaying all
 the roles of the application. This is a usability feature.

### Goals and Non-Goals

Display authorized groups dropdown list only when the user selects the manual judgment stage.
Display all the groups the user has selected while creating application(permissions) in the authorized groups dropdown.
When the pipeline is executed, disable the manual judgment 'continue/cancel' based on the user role selected in the authorized groups.

opsmxemp1 : ["mail-qa","sra-dev","mail-dev","sra-qa","empdev"]

| Testcase No. | Application Name | Application Roles                               | Application Permissions                                             |   Manual Judgment Roles | User      | Result         |
|:------------:|------------------|-------------------------------------------------|---------------------------------------------------------------------|-------------------------|-----------|----------------|
| 1            | testmanualjud001 | sra-dev,sra-qa                                  |sra-dev(r,w,e),sra-qa(r)                                             |                 sra-dev | opsmxemp1 | Enable Button  |
| 2            | testmanualjud002 | N.A                                             |         ----                                                        |                     N.A | opsmxemp1 | Enable Button  |
| 3            | testmanualjud003 | sra-dev,sra-qa                                  |sra-dev(r,w,e),sra-qa(r)                                             |                     N.A | opsmxemp1 | Enable Button  |
| 4            | testmanualjud004 | sra-dev,sra-qa,mail-dev,mail-qa,emp-dev,finance |sra-dev(r),sra-qa(r),mail-dev(r,w,e),mail-qa(r),emp-dev(r),finance(r)| sra-qa,mail-dev,finance | opsmxemp1 | Enable Button  |
| 5            | testmanualjud005 | sra-dev,mail-qa,operations                      |sra-dev(r),mail-qa(r),operations(r,w,e)                              |              operations | opsmxemp1 | Disable Button |
| 6            | testmanualjud006 | sra-dev,mail-qa                                 |sra-dev(r),mail-qa(r,w,e)                                            |              sra-dev    | opsmxemp1 | Disable Button |

User can opt out of this feature as stated in the testcase 2 & 3.
If authz is not enabled, it will work in the same way.
The above test cases will handle all the test scenarios.
If the role is set in the manual judgment stage, then we check the stage role both in the application roles and user roles.
If yes,we check for application permissions of that role. If yes(stage role has execute/write/create application permission),
we can approve and move to the downstream stage, if no(stage role has read application permission), then we deny
(disable the continue button in the manual judgment stage).If the pipeline is triggered from command line,
then the backend orca code checks the same above mentioned logic.If yes,we approve and moves to the
downstream stage , if no, then we deny(It will be running in the same stage).

## Motivation and Rationale

Only legitimate users can approve the manual judgment stage so that it can be moved to downstream stages.

## Timeline

Already Implemented

## Design

1. Enhanced stage.html to

   Get the roles of the application.
   Display the roles of the application as a list only if the stage is Manual Judgment.

2. Enhanced stage.module.js to

   Fetch the permissions of the application from the gate application url.
   Application permissions has all the roles.
   Populate the list with only the roles of the application with no duplicates.

3.  Enhanced ApplicationReader.ts to

   Fetch the permissions of the application from the gate application url.

4.  Enhanced ManualJudgmentApproval.tsx to

   Fetch the roles of the application, stage and the user during the execution.
   Iterate through each of the user role to check if the role exists in the stage and application.
   We also check for stage role application permission.
   If the stage role has execute/write/create application permission, then enable the continue button.
   If the stage role has read application permission, then disable the continue button.

5.  Enhanced OperationsController.groovy to

    Get the application roles, stage roles and the user roles.
    Check each of the user roles whether they are contained in the stage and application roles.
    We also check for stage role application permission.
    If the stage role has execute/write/create application permission, then we set the isAuthorized flag to true.
    If the stage role has read application permission, then we set the isAuthorized flag to false.
    By default, all the stages except Manual Judgment are true.

6.  Enhanced ManualJudgmentStage.groovy to

    Check for the flag in ManualJudgmentStage.groovy whether to execute to the next stage or not.
    If yes/no, continue with the next stages/continues running the same stage.

7.  Enhanced ManualJudgmentStageSpec.groovy to

    Modified the testcases as per the new logic.

ethanfrogers commented:
This RFC is linked to a previous RFC where initial feedback was given - https://github.com/spinnaker/community/pull/78

ajordens commented:
`This is a usability feature` - was there any feedback from the Security SIG? Personally, this seems less focused on usability and more on security/compliance.

Implementation-wise, I'd suggest starting with the restrictions in `orca` and the ability to configure the required set of roles on the Manual Judgment stage. That may or may not uncover an issue with how we currently check auth around `PATCH`'ing a pipeline stage.

The fact that continue/cancel buttons are hidden is definitely a usability improvement but the important checks happen in `orca`.  

As an aside, allowing for free-form group specification (vs only selecting from a list) seems reasonable to me but is an implementation concern. It would be an improvement over `Edit JSON`. 

Overall, I'm 👍 to the idea and see it as an incremental improvement over `Manual Judgment`.

sanopsmx commented:
We did not hear any feedback from security SIG.

Checks happen in both UI and orca.

robzienert commented:
@bpowell @helenelau - Sounds like this RFC is currently in your court.

helenelau commented:
@bpowell let's discuss

helenelau commented:
@sanopsmx can you attend the next security sig meeting on 03/02? 
https://github.com/spinnaker/governance/tree/master/sig-security
https://docs.google.com/document/d/1GhzhmG7ZytJBfW2lontOSBxVghgNZmxstF49SoocTDQ/edit#heading=h.ilm7rdxf4d42

sanopsmx commented:
Yes will attend the meeting on 03/02.

Regards,
Sanjeev

On Thu, 27 Feb 2020 at 02:01, Helene Lau <notifications@github.com> wrote:

> @sanopsmx <https://github.com/sanopsmx> can you attend the next security
> sig meeting on 03/02?
> https://github.com/spinnaker/governance/tree/master/sig-security
>
> https://docs.google.com/document/d/1GhzhmG7ZytJBfW2lontOSBxVghgNZmxstF49SoocTDQ/edit#heading=h.ilm7rdxf4d42
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spinnaker/governance/pull/80?email_source=notifications&email_token=AM4NAFFYBPTHTV7THHEKJZ3RE3GR7A5CNFSM4J2CRKO2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENBYP5I#issuecomment-591628277>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM4NAFCGL23Y2P3X32SAJALRE3GR7ANCNFSM4J2CRKOQ>
> .
>

plumpy commented:
FYI, the security SIG meeting is actually on 3/5.

helenelau commented:
Yes thanks for catching it.  @sanopsmx 03/05 it is.

sanopsmx commented:
Will attend on 03/05. Can you please let us know the time and bridge
details.


On Mon, 2 Mar 2020 at 23:45, Helene Lau <notifications@github.com> wrote:

> Yes thanks for catching it. @sanopsmx <https://github.com/sanopsmx> 03/05
> it is.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spinnaker/governance/pull/80?email_source=notifications&email_token=AM4NAFH3SHOWPMVGWAIRL7LRFPZUZA5CNFSM4J2CRKO2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENQLVFI#issuecomment-593541781>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM4NAFBXBLLGRKB562SEFSTRFPZUZANCNFSM4J2CRKOQ>
> .
>

helenelau commented:
The info is at the links in my previous comment.  In particular, it’s at
the top of the google doc.

On Mon, Mar 2, 2020 at 7:22 PM Sanjeev Thatiparthi <notifications@github.com>
wrote:

> Will attend on 03/05. Can you please let us know the time and bridge
> details.
>
>
> On Mon, 2 Mar 2020 at 23:45, Helene Lau <notifications@github.com> wrote:
>
> > Yes thanks for catching it. @sanopsmx <https://github.com/sanopsmx>
> 03/05
> > it is.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/spinnaker/governance/pull/80?email_source=notifications&email_token=AM4NAFH3SHOWPMVGWAIRL7LRFPZUZA5CNFSM4J2CRKO2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENQLVFI#issuecomment-593541781
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AM4NAFBXBLLGRKB562SEFSTRFPZUZANCNFSM4J2CRKOQ
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spinnaker/governance/pull/80?email_source=notifications&email_token=ACBPO6HD2TD44PEUN4CG5DDRFRZWNA5CNFSM4J2CRKO2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENR5ZRI#issuecomment-593747141>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACBPO6CSQAGTI3JDZQSHJR3RFRZWNANCNFSM4J2CRKOQ>
> .
>

justinrlee commented:
Hi all.  I've (internally) been working on a similar proposal, which has more of a higher-level goal discussion and less technical details - would be nice if I could get some feedback on this, and maybe somebody could work to combine the two proposals or something.

https://docs.google.com/document/d/15zlA80xCrkCn4L08YbrBDDPJsL--9lUlNeWooZy4Nk8/edit?usp=sharing=

helenelau commented:
@justinrlee please join our meeting at 11am pt ... meet.google.com/rhf-kyvh-dre @sanopsmx fyi

justinrlee commented:
Hey @helenelau I was actually hoping to join but I have another prior engagement at that time.  One of my coworkers will be joining to represent in my stead.

robzienert commented:
@helenelau Is there any update from the prior Security SIG here? Was this RFC discussed at all?

sanopsmx commented:
discussed ...Are we having the sig meeting again on march 19th....

On Wed, 18 Mar 2020 at 01:30, Rob Zienert <notifications@github.com> wrote:

> @helenelau <https://github.com/helenelau> Is there any update from the
> prior Security SIG here? Was this RFC discussed at all?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spinnaker/governance/pull/80#issuecomment-600269676>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM4NAFBYR4IM3VJ2S2IDF4DRH7JGZANCNFSM4J2CRKOQ>
> .
>

helenelau commented:
@sanopsmx yes tomorrow at 11am pt ... @justinrlee fyi

robzienert commented:
@sanopsmx It looks like you moved the RFC template. Can you fix the PR so that the template is still there?

sanopsmx commented:
Raised a new pull request

sanopsmx commented:
created a new PR

https://github.com/spinnaker/governance/pull/152

sanopsmx commented:
reopening this PR. Will update the PR.

sanopsmx commented:
Updated the comments...Modified few testcases...incorporated the sig meeting feedback..

helenelau commented:
@bcornils fyi

sanopsmx commented:
Can you share the meeting invite. Is the meeting scheduled on Julu 23rd.

bcornils commented:
Yes. This Thursday at 11 pdt. The link is in the security sig slack channel. 

Sent from my iPhone

> On Jul 21, 2020, at 9:56 PM, Sanjeev Thatiparthi <notifications@github.com> wrote:
> 
> ﻿
> Can you share the meeting invite. Is the meeting scheduled on Julu 23rd.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

maggieneterval commented:
@bcornils @sanopsmx Did the Security SIG accept this RFC? Wanted to check since the related [Deck PR](https://github.com/spinnaker/deck/pull/8368) has been open for several weeks, which I am happy to help review, but wanted to check first to see if the Security SIG wanted further discussion on this proposal.

helenelau commented:
I believe so.   We had a demo back in March where we requested additional
work.  We had another demo during the last sig meeting.  Next step would be
code review.  I have not been able to recruit one yet.

On Mon, Aug 3, 2020 at 1:54 PM Maggie Neterval <notifications@github.com>
wrote:

> @bcornils
> <https://urldefense.com/v3/__https://github.com/bcornils__;!!DCbAVzZNrAf4!XnNK1gUhXMHAozIB1neAbAFbGYQYpS9QUyXcmKXg6wrgmGzCzOBNeelLAiiCEMo$>
> @sanopsmx
> <https://urldefense.com/v3/__https://github.com/sanopsmx__;!!DCbAVzZNrAf4!XnNK1gUhXMHAozIB1neAbAFbGYQYpS9QUyXcmKXg6wrgmGzCzOBNeelLhlJocRI$>
> Did the Security SIG accept this RFC? Wanted to check since the related Deck
> PR
> <https://urldefense.com/v3/__https://github.com/spinnaker/deck/pull/8368__;!!DCbAVzZNrAf4!XnNK1gUhXMHAozIB1neAbAFbGYQYpS9QUyXcmKXg6wrgmGzCzOBNeelLkwfpJNQ$>
> has been open for several weeks, which I am happy to help review, but
> wanted to check first to see if the Security SIG wanted further discussion
> on this proposal.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://urldefense.com/v3/__https://github.com/spinnaker/governance/pull/80*issuecomment-668236656__;Iw!!DCbAVzZNrAf4!XnNK1gUhXMHAozIB1neAbAFbGYQYpS9QUyXcmKXg6wrgmGzCzOBNeelLHpyXcWc$>,
> or unsubscribe
> <https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ACBPO6GEKPII7QZDET2PEZTR64PZJANCNFSM4J2CRKOQ__;!!DCbAVzZNrAf4!XnNK1gUhXMHAozIB1neAbAFbGYQYpS9QUyXcmKXg6wrgmGzCzOBNeelL_ydrYA8$>
> .
>

sanopsmx commented:
@maggieneterval ...I gave the presentation to the sig meeting on July 23rd. Incorporated all the comments suggested in the earlier march meeting. I think, only merge is pending.

maggieneterval commented:
Thanks for the update, @sanopsmx. Should we update this RFC as accepted and change the SIG designation to Security? @helenelau @bcornils I've reviewed the Deck component of this change, but I'll leave it to y'all to identify an appropriate reviewer for the Orca component.

helenelau commented:
Thanks for your help ...

On Tue, Aug 4, 2020 at 8:56 AM Maggie Neterval <notifications@github.com>
wrote:

> Thanks for the update, @sanopsmx
> <https://urldefense.com/v3/__https://github.com/sanopsmx__;!!DCbAVzZNrAf4!QGL0ro9p_y4mXYvQvNIessu1V7l8Gj1ivoRRFfYopCnbcwn10892N3yD046CQpg$>.
> Should we update this RFC as accepted and change the SIG designation to
> Security? @helenelau
> <https://urldefense.com/v3/__https://github.com/helenelau__;!!DCbAVzZNrAf4!QGL0ro9p_y4mXYvQvNIessu1V7l8Gj1ivoRRFfYopCnbcwn10892N3yDvGgYTcU$>
> @bcornils
> <https://urldefense.com/v3/__https://github.com/bcornils__;!!DCbAVzZNrAf4!QGL0ro9p_y4mXYvQvNIessu1V7l8Gj1ivoRRFfYopCnbcwn10892N3yDo_1AVh0$>
> I've reviewed the Deck component of this change, but I'll leave it to y'all
> to identify an appropriate reviewer for the Orca component.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://urldefense.com/v3/__https://github.com/spinnaker/governance/pull/80*issuecomment-668679911__;Iw!!DCbAVzZNrAf4!QGL0ro9p_y4mXYvQvNIessu1V7l8Gj1ivoRRFfYopCnbcwn10892N3yDhgSm9Lo$>,
> or unsubscribe
> <https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ACBPO6ANTCE7CNUMQSDBUQTR7AVR7ANCNFSM4J2CRKOQ__;!!DCbAVzZNrAf4!QGL0ro9p_y4mXYvQvNIessu1V7l8Gj1ivoRRFfYopCnbcwn10892N3yDwLuvuwc$>
> .
>

link108 commented on rfc/.manualjudgment.md:
```diff
@@ -0,0 +1,94 @@
+| **Status**     | _**Proposed**, Accepted, Implemented, Obsolete_ |
+| **RFC #**      | https://github.com/spinnaker/deck/pull/8368 | https://github.com/spinnaker/orca/pull/3763
+| **Author(s)**  | Sanjeev Thatiparthi (@sanopsmx) (https://github.com/sanopsmx)
+| **SIG / WG**   | sig-ux
```
please update to sig-security

link108 commented:
I like the idea of this! I think the Orca PR needs some improvement, but I can give that feedback in the actual PR. One thing I did notice that is relevant to call out here - the PR checks for [isTest](https://github.com/spinnaker/orca/pull/3763/files#diff-78188443fe2868784fbaf81f735334a2R132), which is not noted in this design doc. At first glance, I don't like pushing that logic (checking if an environment is a test environment or not) into a manual judgement stage, but I could be convinced otherwise.

robzienert reviewed (approved):
Approved with update to `sig-security`.

sanopsmx commented:
> I like the idea of this! I think the Orca PR needs some improvement, but I can give that feedback in the actual PR. One thing I did notice that is relevant to call out here - the PR checks for [isTest](https://github.com/spinnaker/orca/pull/3763/files#diff-78188443fe2868784fbaf81f735334a2R132), which is not noted in this design doc. At first glance, I don't like pushing that logic (checking if an environment is a test environment or not) into a manual judgement stage, but I could be convinced otherwise.



> I like the idea of this! I think the Orca PR needs some improvement, but I can give that feedback in the actual PR. One thing I did notice that is relevant to call out here - the PR checks for [isTest](https://github.com/spinnaker/orca/pull/3763/files#diff-78188443fe2868784fbaf81f735334a2R132), which is not noted in this design doc. At first glance, I don't like pushing that logic (checking if an environment is a test environment or not) into a manual judgement stage, but I could be convinced otherwise.

Removed 'isTest' flag according to the comments mentioned.

sanopsmx commented:
Can you review this PR and merge.

Accommodated all the suggested changes in the sig-meeting.

robzienert commented:
@sanopsmx I will merge as soon as you un-delete the RFC template file.

sanopsmx commented:
Renamed the .manualJudgment.md to manualJudgment.md
Added .templated.md

sanopsmx commented:
> @sanopsmx I will merge as soon as you un-delete the RFC template file and rename the RFC to `manualjudgement.md` (remove the leading dot from the filename).

Renamed the .manualJudgment.md to manualJudgment.md
Added .templated.md

sanopsmx commented:
> @sanopsmx I will merge as soon as you un-delete the RFC template file and rename the RFC to `manualjudgement.md` (remove the leading dot from the filename).

@robzienert 

Renamed the .manualJudgment.md to manualJudgment.md
Added .templated.md
